### PR TITLE
Feat: event factory support + from_json helper

### PR DESCRIPTION
**Pull request by kares**
### What does this PR do?

Provides a compatibility layer for LS' `Event.from_json(json) { |data| ... }` by having a `events_from_json` helper.
To be used by plugins such as json codec to implement `target => [here]` support.

### Checklist

- [x] core support from LS https://github.com/elastic/logstash/pull/13017
- [x] tests
- [x] perf.test (confirm no regression against base-line `from_json` performance)

LS 7.13.1
```
Logstash: 7.13.0 jruby 9.2.16.0 (2.5.7) 2021-03-03 f82228dc32 Java HotSpot(TM) 64-Bit Server VM 25.171-b11 on 1.8.0_171-b11 +jit [linux-x86_64]

json_small (pretty) string size: 287

Calculating -------------------------------------
json_small (pretty) from_json       
                        332.737k (± 7.7%) i/s -      9.904M in  30.003951s


regular string size: 6441

Calculating -------------------------------------
regular from_json       
                         26.432k (± 5.3%) i/s -    790.514k in  30.029202s


medium_array string size: 8659

Calculating -------------------------------------
medium_array from_json       
                         13.503k (± 5.5%) i/s -    404.378k in  30.079094s


big_array string size: 13698

Calculating -------------------------------------
big_array from_json       
                          8.106k (± 6.9%) i/s -    242.250k in  30.081600s
                          

huge_array string size: 186924

Calculating -------------------------------------
huge_array from_json       
                        967.864  (± 5.4%) i/s -     29.003k in  30.089901s

```

LS 8.0 (with `from_json {}` support)
```
Logstash: 8.0.0 jruby 9.2.16.0 (2.5.7) 2021-03-03 f82228dc32 Java HotSpot(TM) 64-Bit Server VM 25.171-b11 on 1.8.0_171-b11 +jit [linux-x86_64]

json_small (pretty) string size: 287

Calculating -------------------------------------
json_small (pretty) from_json       
                        335.784k (± 7.2%) i/s -     10.016M in  30.028366s
json_small (pretty) from_json { ... }
                        298.067k (± 6.2%) i/s -      8.918M in  30.072253s

Comparison:
json_small (pretty) from_json       :   335784.5 i/s
json_small (pretty) from_json { ... }:   298066.8 i/s - same-ish: difference falls within error


regular string size: 6441

Calculating -------------------------------------
regular from_json       
                         26.228k (± 7.2%) i/s -    783.300k in  30.081367s
regular from_json { ... }
                         25.427k (± 5.2%) i/s -    761.800k in  30.070375s

Comparison:
regular from_json       :    26227.5 i/s
regular from_json { ... }:    25426.6 i/s - same-ish: difference falls within error


medium_array string size: 8659

Calculating -------------------------------------
medium_array from_json       
                         12.747k (± 6.3%) i/s -    381.900k in  30.124069s
medium_array from_json { ... }
                         13.317k (± 6.0%) i/s -    398.240k in  30.067693s

Comparison:
medium_array from_json { ... }:    13317.2 i/s
medium_array from_json       :    12747.2 i/s - same-ish: difference falls within error


big_array string size: 13698

Calculating -------------------------------------
big_array from_json       
                          8.648k (± 4.9%) i/s -    258.687k in  30.017296s
big_array from_json { ... }
                          8.188k (± 7.4%) i/s -    244.530k in  30.091162s

Comparison:
big_array from_json       :     8647.7 i/s
big_array from_json { ... }:     8187.5 i/s - same-ish: difference falls within error


huge_array string size: 186924

Calculating -------------------------------------
huge_array from_json       
                        919.235  (± 5.1%) i/s -     27.550k in  30.082411s
huge_array from_json { ... }
                        904.792  (± 5.2%) i/s -     27.072k in  30.033867s

Comparison:
huge_array from_json       :      919.2 i/s
huge_array from_json { ... }:      904.8 i/s - same-ish: difference falls within error

```

### Sample (pseudo) use-case
```ruby
#require 'logstash/plugin_mixins/ecs_compatibility_support'
#require 'logstash/plugin_mixins/ecs_compatibility_support/target_check'
require 'logstash/plugin_mixins/validator_support/field_reference_validation_adapter'
require 'logstash/plugin_mixins/event_support/event_factory_adapter'
require 'logstash/plugin_mixins/event_support/from_json_helper'

class LogStash::Codecs::JSON < LogStash::Codecs::Base

  #include LogStash::PluginMixins::ECSCompatibilitySupport
  #include LogStash::PluginMixins::ECSCompatibilitySupport::TargetCheck

  extend LogStash::PluginMixins::ValidatorSupport::FieldReferenceValidationAdapter

  include LogStash::PluginMixins::EventSupport::EventFactoryAdapter
  include LogStash::PluginMixins::EventSupport::FromJsonHelper

  config_name "json"
  
  config :target, :validate => :field_reference

  def decode(data, &block)
    parse_json(data, &block)
  end

  private

  def parse_json(json)
    events_from_json(json, targeted_event_factory).each { |event| yield event }
  rescue => e
    @logger.error("JSON parse error, original data now in message field", message: e.message, exception: e.class, data: json)
    yield parse_json_error_event(json)
  end

  def parse_json_error_event(json)
    event_factory.new_event("message" => json, "tags" => ["_jsonparsefailure"])
  end

end
```

### CODE REVIEW

File: lib/logstash/plugin_mixins/event_support/event_factory_adapter.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        # @raise [ArgumentError]
+        # @return [void]
+        def self.included(base)
+          fail(ArgumentError, "`#{base}` must inherit LogStash::Plugin") unless base < LogStash::Plugin
+
+          if defined? LogStash::Plugins::EventFactorySupport
+            core_event_support = LogStash::Plugins::EventFactorySupport
+            fail(ArgumentError, "`#{base}` should include #{core_event_support}") unless base < core_event_support
+          else
+            require 'logstash/plugin_mixins/event_support/event_factory_adapter/fallback_impl'
```
Comment (yaauie) [inline lib/logstash/plugin_mixins/event_support/event_factory_adapter.rb]: Nit: Since Logstash has such a long include path and our many plugins aren't always the best at properly namespacing their files, I prefer `require_relative` for library-internal requires.

```suggestion
            require_relative 'event_factory_adapter/fallback_impl'
```

File: lib/logstash/plugin_mixins/event_support/event_factory_adapter/fallback_impl.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+module LogStash::PluginMixins::EventSupport::EventFactoryAdapter::FallbackImpl
+
+  # The event_factory method is effectively final and should not be re-defined by plugins.
+  #
+  # @return an event factory object with a `new_event(Hash)` API
+  def event_factory
+    @_event_factory || JRuby.reference(self).synchronized do
+      @_event_factory ||= create_event_factory
+    end
```
Comment (yaauie) [inline lib/logstash/plugin_mixins/event_support/event_factory_adapter/fallback_impl.rb]: I _think_ that we can afford to be simpler here in our fallback implementation and just return the `BasicEventFactory` singleton. By definition, if we are using the fallback implementation we aren't able to use core features so it doesn't make sense for core to need to hook into this.

```suggestion
    BasicEventFactory::INSTANCE
```

File: lib/logstash/plugin_mixins/event_support/event_factory_adapter/fallback_impl.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                                 end
+  end
+
+  private
+
+  # @api private
+  # @since LS 7.14
+  def create_event_factory
+    BasicEventFactory::INSTANCE
+  end
```
Comment (yaauie) [inline lib/logstash/plugin_mixins/event_support/event_factory_adapter/fallback_impl.rb]: If we take my other suggestion to simplify `FallbackImpl#event_factory`, we won't need this:

```suggestion
```

File: lib/logstash/plugin_mixins/event_support/from_json_helper.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          def events_from_json(json, event_factory)
+            LogStash::Event.from_json(json) { |data| event_factory.new_event(data) }
+          end
+
+        end
+        private_constant :NativeImpl
+
+        module FallbackImpl
+
+          def events_from_json(json, event_factory)
```
Comment (yaauie) [inline lib/logstash/plugin_mixins/event_support/from_json_helper.rb:34]: Is it worth short-cutting to `Event::from_json` for performance if we are working with a `BasicEventFactory`? Doing so would require that `BasicEventFactory` not be a private constant.

```suggestion
          def events_from_json(json, event_factory)
            return Logstash::Event.from_json(json) if event_factory.kind_of?(BasicEventFactory)
```

File: spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+          describe 'targeted_event_factory' do
+
+            subject(:plugin_class) do
+              Class.new(plugin_base_class) do
+                config_name 'sample'
+                config :target, :validate => :string
+              end
+            end
+
+            subject(:event_factory) { plugin.send(:targeted_event_factory) }
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb]: Nit: This led to a little confusion when reading the specs, since our plugin also has a bare `event_factory`. Perhaps we can rename it to be more clear?

```suggestion
            subject(:targeted_event_factory) { plugin.send(:targeted_event_factory) }
```

File: spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+            it_behaves_like 'an event factory'
+
+            it 'memoizes the factory instance' do
+              expect( event_factory ).to be plugin.send(:targeted_event_factory)
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb]: nit(batch):
```suggestion
              expect( targeted_event_factory ).to be plugin.send(:targeted_event_factory)
```

File: spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            end
+
+            it 'uses the basic event factory (no target specified)' do
+              expect( event_factory ).to be plugin.send(:event_factory)
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb]: nit(batch):

```suggestion
              expect( targeted_event_factory ).to be plugin.send(:event_factory)
```

File: spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            end
+
+            context 'with target' do
+
+              let(:options) { super().merge('target' => '[the][baz]') }
+
+              it 'returns an event' do
+                expect( event_factory.new_event ).to be_a LogStash::Event
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb]: nit(batch):

```suggestion
                expect( targeted_event_factory.new_event ).to be_a LogStash::Event
```

File: spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+                expect( event = event_factory.new_event('foo' => 'bar') ).to be_a LogStash::Event
+                expect( event.include?('foo') ).to be false
+                expect( event.get('[the][baz][foo]') ).to eql 'bar'
+              end
+
+              it 'memoizes the factory instance' do
+                expect( event_factory ).to be plugin.send(:targeted_event_factory)
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb]: nit(batch):

```suggestion
                expect( targeted_event_factory ).to be plugin.send(:targeted_event_factory)
```

File: spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+              end
+
+              it 'uses a different factory fro the basic one' do
+                expect( event_factory ).not_to be plugin.send(:event_factory)
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb]: nit(batch):
```suggestion
                expect( targeted_event_factory ).not_to be plugin.send(:event_factory)
```

File: spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+                expect( event = event_factory.new_event('foo' => 'bar') ).to be_a LogStash::Event
+                expect( event.include?('foo') ).to be false
+                expect( event.get('[the][baz][foo]') ).to eql 'bar'
+              end
+
+              it 'memoizes the factory instance' do
+                expect( event_factory ).to be plugin.send(:targeted_event_factory)
+              end
+
+              it 'uses a different factory fro the basic one' do
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb]: typo

```suggestion
              it 'uses a different factory from the basic one' do
```

File: spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+            subject(:plugin_class) do
+              Class.new(plugin_base_class) do
+                config_name 'sample'
+                config :target, :validate => :string
+              end
+            end
+
+            subject(:event_factory) { plugin.send(:targeted_event_factory) }
+
+            it_behaves_like 'an event factory'
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb]: nit(batch): -- I'm not 100% sure on defining a named `subject` to refer to an outer `subject` by name? An alternative would be to refer to `subject` directly in the shared example group instead of referring to the `event_factory` by name.

```suggestion
            it_behaves_like 'an event factory' do
              subject(:event_factory) { targeted_event_factory }
            end
```

File: spec/logstash/plugin_mixins/event_support/from_json_helper_spec.rb [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        it 'parses multi json' do
+          expect( event_factory ).to receive(:new_event) { |data| LogStash::Event.new('test' => data) }.twice
+
+          events = plugin.events_from_json('[ {"foo": "bar"}, { "baz": 42.0 } ]', event_factory)
+          expect( events.size ).to eql 2
+          expect( events[0].get('[test]') ).to eql 'foo' => 'bar'
+          expect( events[1].get('[test]') ).to eql 'baz' => 42.0
+        end
+
+        it 'raises on unexpecred json' do
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/from_json_helper_spec.rb]: typo

```suggestion
        it 'raises on unexpected json' do
```

File: spec/logstash/plugin_mixins/event_support/from_json_helper_spec.rb [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+          expect { plugin.events_from_json(' "42" ', event_factory) }.to raise_error(LogStash::Json::ParserError)
+        end
+
+        it 'raises on invalid json' do
+          expect { plugin.events_from_json('{ "" }', event_factory) }.to raise_error(LogStash::Json::ParserError)
+        end
```
Comment (yaauie) [inline spec/logstash/plugin_mixins/event_support/from_json_helper_spec.rb:60]: I've been bit by parsers that hang when the parser is in a "good" state and hits EOF, so it'd be good to cover that case too.

```suggestion
        end
        
        it 'raises on incomplete json' do
          expect { plugin.events_from_json('{"answer":"42"', event_factory) }.to raise_error(LogStash::Json::ParserError)
        end
```

**Review by yaauie — COMMENTED** [review]
This is absolutely excellent 🥇 

Structurally it makes a lot of sense and absolutely gets us to the point where we can start using the EventFactory in plugins _and_ bridges the gap to the `Event#from_json` event factory support. Absolutely well done.

I've left what looks like a bloodbath of nitpick suggestions for a single rename and its fallout 😬, along with a couple typos.


Open threads: feel free to ignore:
 - I think we _may_ be able to simplify the `EventFactoryAdapter::FallbackImpl#event_factory` by just using the `BasicEventFactory::INSTANCE` instead of routing through `create_event_factory`
 - In `FromJsonHelper::FallbackImpl#events_from_json` should we fall through to native for performance when we are given a `EventFactoryAdapter::FallbackImpl::BasicEventFactory`, or does that tangle things up too much?

File: lib/logstash/plugin_mixins/event_support/event_factory_adapter/fallback_impl.rb (hunk 2 above)
Comment (kares) [reply lib/logstash/plugin_mixins/event_support/event_factory_adapter/fallback_impl.rb]: okay I did a slightly different variation: `@_event_factory ||= BasicEventFactory::INSTANCE`
which looks weird but still communicates an intent of a factory being lazy initialized (as I think we'll revisit some stateful form of `create_event_factory` in the future). 

on a related note I would also remove `create_event_factory` from core for now, wdyt?

the reasoning being I am not happy with doing a `create_event_factory(config = self.config)` to make it more useful for now due the fact that it treats `config['target']` in a special way ... think we should wait for the use-case when a plugin would need to instantiate an event factory manually.

File: spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb (hunk 12 above)
Comment (kares) [reply spec/logstash/plugin_mixins/event_support/event_factory_adapter_spec.rb]: updated to use `targeted_event_factory` name, thanks for the suggestion(s)

File: lib/logstash/plugin_mixins/event_support/from_json_helper.rb (hunk 4 above)
Comment (kares) [reply lib/logstash/plugin_mixins/event_support/from_json_helper.rb:34]: run a few micro benchmarks and the block vs non-block form only differ on small JSON payloads (< 1kB) e.g.
```
json_small (compact) string size: 195
Warming up --------------------------------------
json_small (compact) from_json       
                        33.432k i/100ms
json_small (compact) from_json { ... }
                        30.001k i/100ms
Calculating -------------------------------------
json_small (compact) from_json       
                        360.141k (± 5.4%) i/s -     10.765M in  30.017677s
json_small (compact) from_json { ... }
                        312.449k (± 7.5%) i/s -      9.300M in  30.013306s


json_small (pretty) string size: 287
Warming up --------------------------------------
json_small (pretty) from_json       
                        29.900k i/100ms
json_small (pretty) from_json { ... }
                        27.106k i/100ms
Calculating -------------------------------------
json_small (pretty) from_json       
                        335.784k (± 7.2%) i/s -     10.016M in  30.028366s
json_small (pretty) from_json { ... }
                        298.067k (± 6.2%) i/s -      8.918M in  30.072253s

Comparison:
json_small (pretty) from_json       :   335784.5 i/s
json_small (pretty) from_json { ... }:   298066.8 i/s - same-ish: difference falls within error

```

as the payload gets bigger it's less noticeable:

```
regular string size: 6441
Warming up --------------------------------------
regular from_json       
                         2.611k i/100ms
regular from_json { ... }
                         2.344k i/100ms
Calculating -------------------------------------
regular from_json       
                         26.228k (± 7.2%) i/s -    783.300k in  30.081367s
regular from_json { ... }
                         25.427k (± 5.2%) i/s -    761.800k in  30.070375s

Comparison:
regular from_json       :    26227.5 i/s
regular from_json { ... }:    25426.6 i/s - same-ish: difference falls within error



medium_array string size: 8659
Warming up --------------------------------------
medium_array from_json       
                         1.273k i/100ms
medium_array from_json { ... }
                         1.310k i/100ms
Calculating -------------------------------------
medium_array from_json       
                         12.747k (± 6.3%) i/s -    381.900k in  30.124069s
medium_array from_json { ... }
                         13.317k (± 6.0%) i/s -    398.240k in  30.067693s

Comparison:
medium_array from_json { ... }:    13317.2 i/s
medium_array from_json       :    12747.2 i/s - same-ish: difference falls within error

```
think we're good here for now

**Review by yaauie — APPROVED** [review]
LGTM :+1:

File: lib/logstash/plugin_mixins/event_support/event_factory_adapter/fallback_impl.rb (hunk 2 above)
Comment (yaauie) [reply lib/logstash/plugin_mixins/event_support/event_factory_adapter/fallback_impl.rb]: Similar topic, but from the LS core perspective

> It's private, so not necessarily something we need to align on. It is also a good hook point for future efforts in LS core that don't need backport support (things like alternate Event implementation experiments like copy-on-write or field reference parser changes), so I expect even if we remove it now we will add it back in short order.

I don't see a world in which plugins ever need to reach in under the hood and change the implementation details of the event factory. This support gem is in place so that plugins can _use_ the event factory APIs on older logstashes where they aren't provided by LS core. I see a future LS core providing alternate implementations based on a pipeline's configuration (such as copy-on-write events, events that use an alternate field reference parser, other things), but which implementation it provides is outside the scope of this support adapter gem.

So in short: LS core was merged with a [_private_ `LogStash::Plugins::EventFactorySupport#create_event_factory`](https://github.com/elastic/logstash/pull/13017/files#diff-b85d5060885a699db3d00ac4d5c77e33edeecf9540aa1f267f2a467f68fa8afeR30), but that is an implementation detail of LS core and there is no need to mirror this in the support adapter's implementation.
